### PR TITLE
Go 1.7 context

### PR DESCRIPTION
**Pull request by ejholmes**
Related https://github.com/remind101/pkg/issues/40

This is a WIP branch to add Go 1.7 support and use package `context` in the stdlib. I don't really expect this to actually get merged anytime soon, but opening a PR so we can comment on it.

The biggest change is removing the `httpx.Handler` interface, and changing everything to implement `http.Handler`. The way I see it, we have 2 choices:
1. Change the signature of httpx.Handler to:
   
   ``` diff
   - ServeHTTPContext(ctx context.Context, w http.ResponseWriter, r *http.Request) error
   + ServeHTTPContext(w http.ResponseWriter, r *http.Request) error
   ```
2. Or just remove it.

I'm in favor of just removing it and encouraging usage of http.Handler, since it's more portable across libraries in the Go ecosystem. It's fairly easy for consumers to [add their own type that returns an error](https://gist.github.com/ejholmes/847e635f6d8f83950232e1ac84365621) (in fact, that's what Empire's http server does in Go 1.7). Doing this also means we no longer need to wrap gorilla/mux.

There's also some backwards incompatible behavior changes:
- `middleware.NewRelicTracer`, and `middelware.NewResponseTimeReporter` requires that the middleware be added after route dispatch has happened by gorilla mux. This is because, we were originally relying on some hacks in httpx.Router to support pulling out the route template. But it's quite a bit easier/cleaner if we just rely on mux.CurrentRoute. This change also opens it up to support other routes more easily. Again, I think this is easy to deal with by encouraging a design like [this](https://gist.github.com/ejholmes/847e635f6d8f83950232e1ac84365621).

/cc @v-yarotsky @sanjayprabhu

### CODE REVIEW

**Comment by v-yarotsky** [comment]
I have a question that is not directly related to this.
Say, we have a main router that mounts and unauthed endpoints (like /health) and authed endpoints (another router).
How hard would it be to obtain the full route template for nested routers?

File: httpx/middleware/metrics.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	t.SetTags(map[string]string{
+		"route":  route,
+		"status": status,
+	})
+}
+
+func templatePath(r *http.Request) string {
+	path := "unknown"
+
+	if route := mux.CurrentRoute(r); route != nil {
```
Comment (v-yarotsky) [inline httpx/middleware/metrics.go]: is it possible to shove current route template into request context upon dispatching?

**Comment by ejholmes** [comment]
Do you mean, something like this?


```go
package main

import (
	"fmt"
	"net/http"
	"net/http/httptest"

	"github.com/gorilla/mux"
)

func handler(w http.ResponseWriter, r *http.Request) {
	route := mux.CurrentRoute(r)
	fmt.Println(route.GetPathTemplate())
}

func main() {
	r1 := mux.NewRouter()

	r2 := r1.PathPrefix("/path").Subrouter()
	r2.HandleFunc("/nested", handler)

	r, _ := http.NewRequest("GET", "/path/nested", nil)
	w := httptest.NewRecorder()

	r1.ServeHTTP(w, r)
}

```

In that case, `route.GetPathTemplate()` is going to return `/path/nested`.

**Comment by v-yarotsky** [comment]
@ejholmes nice!

File: .travis.yml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 language: go
 
 go:
-  - 1.4
+  - tip
```
Comment (sanjayprabhu) [inline .travis.yml]: Why `tip`?

**Pull request closed without merging**